// docpad/out/lib/models/document.js
// Generated by CoffeeScript 1.4.0
var CSON, DocumentModel, FileModel, YAML, balUtil, mime, pathUtil, _,
  __hasProp = {}.hasOwnProperty,
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
  __indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

pathUtil = require('path');

balUtil = require('bal-util');

_ = require('underscore');

mime = require('mime');

CSON = null;

YAML = null;

FileModel = require(__dirname + '/file');

DocumentModel = (function(_super) {

  __extends(DocumentModel, _super);

  function DocumentModel() {
    return DocumentModel.__super__.constructor.apply(this, arguments);
  }

  DocumentModel.prototype.type = 'document';

  DocumentModel.prototype.defaults = {
    referencesOthers: false,
    header: null,
    parser: null,
    body: null,
    rendered: false,
    contentRendered: null,
    contentRenderedWithoutLayouts: null,
    dynamic: false,
    tags: null,
    renderSingleExtensions: false
  };

  DocumentModel.prototype.getOutContent = function() {
    var content;
    content = this.get('contentRendered') || this.get('content') || this.getBuffer();
    return content;
  };

  DocumentModel.prototype.toJSON = function() {
    var data;
    data = DocumentModel.__super__.toJSON.apply(this, arguments);
    data.meta = this.getMeta().toJSON();
    return data;
  };

  DocumentModel.prototype.referencesOthers = function(flag) {
    if (flag == null) {
      flag = true;
    }
    this.set({
      referencesOthers: flag
    });
    return this;
  };

  DocumentModel.prototype.parse = function(opts, next) {
    var buffer, key, meta, reset, value, _ref, _ref1,
      _this = this;
    if (opts == null) {
      opts = {};
    }
    _ref = this.getActionArgs(opts, next), opts = _ref.opts, next = _ref.next;
    buffer = this.getBuffer();
    meta = this.getMeta();
    reset = {};
    _ref1 = meta.attributes;
    for (key in _ref1) {
      if (!__hasProp.call(_ref1, key)) continue;
      value = _ref1[key];
      reset[key] = this.defaults[key];
    }
    reset = balUtil.dereference(reset);
    this.set(reset);
    this.layout = null;
    meta.clear();
    DocumentModel.__super__.parse.call(this, buffer, function() {
      var body, content, err, header, ignored, match, metaData, metaDate, metaUrl, metaUrls, parser, regex, seperator;
      content = _this.get('content');
      regex = /^\s*(([^\s\d\w])\2{2,})(?:\x20*([a-z]+))?([\s\S]*?)\1/;
      match = regex.exec(content);
      if (match) {
        seperator = match[1];
        parser = match[3] || 'yaml';
        header = match[4].trim();
        body = content.substring(match[0].length).trim();
        try {
          switch (parser) {
            case 'cson':
            case 'coffee':
            case 'coffeescript':
            case 'coffee-script':
              if (!CSON) {
                CSON = require('cson');
              }
              metaData = CSON.parseSync(header);
              meta.set(metaData);
              break;
            case 'yaml':
              if (!YAML) {
                YAML = require('yamljs');
              }
              metaData = YAML.parse(header);
              meta.set(metaData);
              break;
            default:
              err = new Error("Unknown meta parser: " + parser);
              return next(err);
          }
        } catch (err) {
          return next(err);
        }
      } else {
        body = content;
      }
      body = body.replace(/^\n+/, '');
      _this.set({
        source: content,
        content: body,
        header: header,
        body: body,
        parser: parser,
        name: _this.get('name') || _this.get('title') || _this.get('basename')
      });
      metaDate = meta.get('date');
      if (metaDate) {
        metaDate = new Date(metaDate);
        meta.set({
          date: metaDate
        });
      }
      ignored = meta.get('ignored') || meta.get('ignore') || meta.get('skip') || meta.get('draft') || (meta.get('published') === false);
      if (ignored) {
        meta.set({
          ignored: true
        });
      }
      metaUrls = meta.get('urls');
      metaUrl = meta.get('url');
      if (metaUrls) {
        _this.addUrl(metaUrls);
      }
      if (metaUrl) {
        _this.addUrl(metaUrl);
      }
      _this.set(meta.toJSON());
      return next();
    });
    return this;
  };

  DocumentModel.prototype.normalize = function(opts, next) {
    var _ref,
      _this = this;
    if (opts == null) {
      opts = {};
    }
    _ref = this.getActionArgs(opts, next), opts = _ref.opts, next = _ref.next;
    DocumentModel.__super__.normalize.call(this, function() {
      var extensions, outExtension;
      extensions = _this.get('extensions');
      if ((extensions != null) && extensions.length) {
        outExtension = extensions[0];
        _this.set({
          outExtension: outExtension
        });
      }
      return next();
    });
    return this;
  };

  DocumentModel.prototype.contextualize = function(opts, next) {
    var _ref,
      _this = this;
    if (opts == null) {
      opts = {};
    }
    _ref = this.getActionArgs(opts, next), opts = _ref.opts, next = _ref.next;
    DocumentModel.__super__.contextualize.call(this, function() {
      return _this.getEve(function(err, eve) {
        var basename, changes, extensions, fullPath, meta, name, outContentType, outExtension, outFilename, outPath, relativeDirPath, relativeOutPath, url;
        if (err) {
          return next(err);
        }
        changes = {};
        meta = _this.getMeta();
        fullPath = _this.get('fullPath');
        basename = _this.get('basename');
        relativeDirPath = _this.get('relativeDirPath');
        extensions = _this.get('extensions');
        outExtension = _this.get('outExtension');
        url = meta.get('url') || null;
        name = meta.get('name') || null;
        outPath = meta.get('outPath') || null;
        outFilename = null;
        if (eve != null) {
          outExtension = eve.get('outExtension');
        }
        if (basename && outExtension) {
          if (basename[0] === '.' && outExtension === extensions[0]) {
            outFilename = basename;
          } else {
            outFilename = "" + basename + "." + outExtension;
          }
          changes.outFilename = outFilename;
        }
        if (outFilename) {
          if (relativeDirPath) {
            relativeOutPath = "" + relativeDirPath + "/" + outFilename;
          } else {
            relativeOutPath = "" + outFilename;
          }
          changes.relativeOutPath = relativeOutPath;
          if (!url) {
            changes.url = url = "/" + relativeOutPath;
          }
        }
        if (!name && (outFilename != null)) {
          changes.name = name = outFilename;
        }
        if (_this.outDirPath) {
          changes.outPath = outPath = pathUtil.join(_this.outDirPath, relativeOutPath);
        }
        if (url) {
          _this.removeUrl(_this.get('url'));
          _this.setUrl(url);
        }
        if (outPath || fullPath) {
          changes.outContentType = outContentType = mime.lookup(outPath || fullPath);
        }
        _this.set(changes);
        return next();
      });
    });
    return this;
  };

  DocumentModel.prototype.hasLayout = function() {
    return this.get('layout') != null;
  };

  DocumentModel.prototype.getLayout = function(next) {
    var file, layoutId;
    file = this;
    layoutId = this.get('layout');
    if (!layoutId) {
      return next(null, null);
    } else if (this.layout && layoutId === this.layout.id) {
      return next(null, this.layout);
    } else {
      this.emit('getLayout', {
        layoutId: layoutId
      }, function(err, opts) {
        var layout;
        layout = opts.layout;
        if (err) {
          return next(err);
        } else if (!layout) {
          err = new Error("Could not find the specified layout: " + layoutId);
          return next(err);
        } else {
          file.set({
            'layout': layout.id
          });
          file.layout = layout;
          return next(null, layout);
        }
      });
    }
    return this;
  };

  DocumentModel.prototype.getEve = function(next) {
    if (this.hasLayout()) {
      this.getLayout(function(err, layout) {
        if (err) {
          return next(err, null);
        } else {
          return layout.getEve(next);
        }
      });
    } else {
      next(null, this);
    }
    return this;
  };

  DocumentModel.prototype.renderExtensions = function(opts, next) {
    var content, context, extension, extensions, extensionsReversed, file, filename, index, renderSingleExtensions, result, tasks, templateData, _i, _j, _len, _len1, _ref;
    file = this;
    extensions = this.get('extensions');
    filename = this.get('filename');
    content = opts.content, templateData = opts.templateData, renderSingleExtensions = opts.renderSingleExtensions;
    if (content == null) {
      content = this.get('body');
    }
    if (templateData == null) {
      templateData = {};
    }
    if (renderSingleExtensions == null) {
      renderSingleExtensions = this.get('renderSingleExtensions');
    }
    result = content;
    extensionsReversed = [];
    if (extensions.length === 0 && filename) {
      extensionsReversed.push(filename);
    }
    for (_i = 0, _len = extensions.length; _i < _len; _i++) {
      extension = extensions[_i];
      extensionsReversed.unshift(extension);
    }
    if (renderSingleExtensions && extensionsReversed.length === 1) {
      if (renderSingleExtensions !== 'auto' || filename.replace(/^\./, '') === extensionsReversed[0]) {
        extensionsReversed.push(null);
      }
    }
    if (extensionsReversed.length <= 1) {
      return next(null, result);
    }
    tasks = new balUtil.Group(function(err) {
      return next(err, result);
    });
    _ref = extensionsReversed.slice(1);
    for (index = _j = 0, _len1 = _ref.length; _j < _len1; index = ++_j) {
      extension = _ref[index];
      context = {
        inExtension: extensionsReversed[index],
        outExtension: extension
      };
      tasks.push(context, function(complete) {
        var eventData;
        eventData = {
          inExtension: this.inExtension,
          outExtension: this.outExtension,
          templateData: templateData,
          file: file,
          content: result
        };
        return file.trigger('render', eventData, function(err) {
          var message;
          if (err) {
            return complete(err);
          }
          if (result && result === eventData.content) {
            message = "\n  Rendering the extension \"" + eventData.inExtension + "\" to \"" + eventData.outExtension + "\" on \"" + (file.attributes.relativePath || file.attributes.fullPath) + "\" didn't do anything.\n  Explanation here: http://docpad.org/extension-not-rendering";
            file.log('warn', message);
            return complete();
          }
          result = eventData.content;
          return complete();
        });
      });
    }
    tasks.sync();
    return this;
  };

  DocumentModel.prototype.renderDocument = function(opts, next) {
    var content, eventData, extension, file, templateData, _ref;
    file = this;
    extension = this.get('extensions')[0];
    _ref = this.getActionArgs(opts, next), opts = _ref.opts, next = _ref.next;
    content = opts.content, templateData = opts.templateData;
    if (content == null) {
      content = this.get('body');
    }
    if (templateData == null) {
      templateData = {};
    }
    eventData = {
      extension: extension,
      templateData: templateData,
      file: file,
      content: content
    };
    file.trigger('renderDocument', eventData, function(err) {
      return next(err, eventData.content);
    });
    return this;
  };

  DocumentModel.prototype.renderLayouts = function(opts, next) {
    var content, file, templateData, _ref;
    file = this;
    _ref = this.getActionArgs(opts, next), opts = _ref.opts, next = _ref.next;
    content = opts.content, templateData = opts.templateData;
    if (content == null) {
      content = this.get('body');
    }
    if (templateData == null) {
      templateData = {};
    }
    return file.getLayout(function(err, layout) {
      if (err) {
        return next(err, content);
      }
      if (layout) {
        templateData.content = content;
        return layout.render({
          templateData: templateData
        }, function(err, result) {
          return next(err, result);
        });
      } else {
        return next(null, content);
      }
    });
  };

  DocumentModel.prototype.render = function(opts, next) {
    var contentRenderedWithoutLayouts, file, fullPath, tasks, _base, _base1, _ref, _ref1, _ref2, _ref3, _ref4;
    if (opts == null) {
      opts = {};
    }
    file = this;
    contentRenderedWithoutLayouts = null;
    fullPath = this.get('fullPath');
    _ref = this.getActionArgs(opts, next), opts = _ref.opts, next = _ref.next;
    opts = _.clone(opts || {});
    if ((_ref1 = opts.actions) == null) {
      opts.actions = ['renderExtensions', 'renderDocument', 'renderLayouts'];
    }
    if ((_ref2 = opts.content) == null) {
      opts.content = this.get('body');
    }
    opts.templateData = _.clone(opts.templateData || {});
    if ((_ref3 = (_base = opts.templateData).document) == null) {
      _base.document = file.toJSON();
    }
    if ((_ref4 = (_base1 = opts.templateData).documentModel) == null) {
      _base1.documentModel = file;
    }
    file.log('debug', "Rendering the file: " + fullPath);
    tasks = new balUtil.Group(function(err) {
      var contentRendered, rendered;
      if (err) {
        file.log('warn', "Something went wrong while rendering: " + fullPath);
        return next(err, opts.content, file);
      }
      contentRendered = opts.content;
      if (contentRenderedWithoutLayouts == null) {
        contentRenderedWithoutLayouts = contentRendered;
      }
      rendered = true;
      file.set({
        contentRendered: contentRendered,
        contentRenderedWithoutLayouts: contentRenderedWithoutLayouts,
        rendered: rendered
      });
      file.log('debug', "Rendering completed for: " + fullPath);
      return next(null, opts.content, file);
    });
    if (__indexOf.call(opts.actions, 'renderExtensions') >= 0) {
      tasks.push(function(complete) {
        return file.renderExtensions(opts, function(err, result) {
          if (err) {
            return complete(err);
          }
          opts.content = result;
          return complete();
        });
      });
    }
    if (__indexOf.call(opts.actions, 'renderDocument') >= 0) {
      tasks.push(function(complete) {
        return file.renderDocument(opts, function(err, result) {
          if (err) {
            return complete(err);
          }
          opts.content = result;
          contentRenderedWithoutLayouts = result;
          return complete();
        });
      });
    }
    if (__indexOf.call(opts.actions, 'renderLayouts') >= 0) {
      tasks.push(function(complete) {
        return file.renderLayouts(opts, function(err, result) {
          if (err) {
            return complete(err);
          }
          opts.content = result;
          return complete();
        });
      });
    }
    tasks.sync();
    return this;
  };

  DocumentModel.prototype.writeRendered = function(next) {
    var encoding, file, fileOutPath, outContent;
    file = this;
    fileOutPath = this.get('outPath');
    encoding = this.get('encoding');
    outContent = this.getOutContent();
    file.log('debug', "Writing the rendered file: " + fileOutPath + " " + encoding);
    balUtil.writeFile(fileOutPath, outContent, function(err) {
      if (err) {
        return next(err);
      }
      file.log('debug', "Wrote the rendered file: " + fileOutPath + " " + encoding);
      return next();
    });
    return this;
  };

  DocumentModel.prototype.writeSource = function(next) {
    var body, content, file, fullPath, header, meta, metaData, parser, seperator, source;
    file = this;
    meta = this.getMeta();
    if (!CSON) {
      CSON = require('cson');
    }
    fullPath = this.get('fullPath');
    content = (this.getContent() || '').toString();
    parser = 'cson';
    seperator = '---';
    file.log('debug', "Writing the source file: " + fullPath);
    metaData = meta.toJSON();
    console.log({
      metaData: metaData
    });
    header = CSON.stringifySync(metaData);
    content = body = content.replace(/^\s+/, '');
    source = "" + seperator + " " + parser + "\n" + header + "\n" + seperator + "\n\n" + body;
    this.set({
      parser: parser,
      header: header,
      body: body,
      content: content,
      source: source
    });
    balUtil.writeFile(fullPath, source, function(err) {
      if (err) {
        return next(err);
      }
      file.log('info', "Wrote the source file: " + fullPath);
      return next();
    });
    return this;
  };

  return DocumentModel;

})(FileModel);

module.exports = DocumentModel;
