// docpad/out/lib/interfaces/console.js
// Generated by CoffeeScript 1.4.0
var ConsoleInterface, cliColor, pathUtil,
  __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
  __hasProp = {}.hasOwnProperty;

cliColor = require('caterpillar').cliColor;

pathUtil = require('path');

ConsoleInterface = (function() {

  function ConsoleInterface(opts, next) {
    this.watch = __bind(this.watch, this);

    this.skeleton = __bind(this.skeleton, this);

    this.clean = __bind(this.clean, this);

    this.server = __bind(this.server, this);

    this.run = __bind(this.run, this);

    this.render = __bind(this.render, this);

    this.install = __bind(this.install, this);

    this.info = __bind(this.info, this);

    this.help = __bind(this.help, this);

    this.generate = __bind(this.generate, this);

    this.welcomeCallback = __bind(this.welcomeCallback, this);

    this.selectSkeletonCallback = __bind(this.selectSkeletonCallback, this);

    this.extractConfig = __bind(this.extractConfig, this);

    this.completeAction = __bind(this.completeAction, this);

    this.performAction = __bind(this.performAction, this);

    this.wrapAction = __bind(this.wrapAction, this);

    this.handleError = __bind(this.handleError, this);

    this.getCommander = __bind(this.getCommander, this);

    this.start = __bind(this.start, this);

    var commander, consoleInterface, docpad, locale, version;
    consoleInterface = this;
    this.docpad = docpad = opts.docpad;
    this.commander = commander = require('commander');
    locale = docpad.getLocale();
    version = require(__dirname + '/../../../package.json').version;
    commander.version(version).option('-o, --out <outPath>', locale.consoleOptionOut).option('-c, --config <configPath>', locale.consoleOptionConfig).option('-e, --env <environment>', locale.consoleOptionEnv).option('-d, --debug [logLevel]', locale.consoleOptionDebug, parseInt).option('-f, --force', locale.consoleOptionForce);
    commander.command('run').description(locale.consoleDescriptionRun).option('-s, --skeleton <skeleton>', locale.consoleOptionSkeleton).option('-p, --port <port>', locale.consoleOptionPort, parseInt).action(consoleInterface.wrapAction(consoleInterface.run));
    commander.command('server').description(locale.consoleDescriptionServer).option('-p, --port <port>', locale.consoleOptionPort, parseInt).action(consoleInterface.wrapAction(consoleInterface.server));
    commander.command('skeleton').description(locale.consoleDescriptionSkeleton).option('-s, --skeleton <skeleton>', locale.consoleOptionSkeleton).action(consoleInterface.wrapAction(consoleInterface.skeleton));
    commander.command('render [path]').description(locale.consoleDescriptionRender).action(function(command) {
      var _ref;
      if ((_ref = commander.debug) == null) {
        commander.debug = 5;
      }
      commander.checkVersion = false;
      commander.welcome = false;
      commander.prompts = false;
      return consoleInterface.performAction(command, consoleInterface.render);
    });
    commander.command('generate').description(locale.consoleDescriptionGenerate).action(consoleInterface.wrapAction(consoleInterface.generate));
    commander.command('watch').description(locale.consoleDescriptionWatch).action(consoleInterface.wrapAction(consoleInterface.watch));
    commander.command('install').description(locale.consoleDescriptionInstall).action(consoleInterface.wrapAction(consoleInterface.install));
    commander.command('clean').description(locale.consoleDescriptionClean).action(consoleInterface.wrapAction(consoleInterface.clean));
    commander.command('info').description(locale.consoleDescriptionInfo).action(consoleInterface.wrapAction(consoleInterface.info));
    commander.command('help').description(locale.consoleDescriptionHelp).action(consoleInterface.wrapAction(consoleInterface.help));
    commander.command('*').description(locale.consoleDescriptionUnknown).action(function() {
      return commander.emit('help', []);
    });
    docpad.on('welcome', function(data, next) {
      return consoleInterface.welcomeCallback(data, next);
    });
    docpad.emitSync('consoleSetup', {
      consoleInterface: consoleInterface,
      commander: commander
    }, function(err) {
      if (err) {
        return consoleInterface.handleError(err);
      }
      return next(null, consoleInterface);
    });
    this;

  }

  ConsoleInterface.prototype.start = function(argv) {
    this.commander.parse(argv || process.argv);
    return this;
  };

  ConsoleInterface.prototype.getCommander = function() {
    return this.commander;
  };

  ConsoleInterface.prototype.handleError = function(err) {
    var docpad, locale;
    docpad = this.docpad;
    locale = docpad.getLocale();
    docpad.log('error', locale.consoleError);
    docpad.error(err);
    process.exit(1);
    return this;
  };

  ConsoleInterface.prototype.wrapAction = function(action) {
    var consoleInterface;
    consoleInterface = this;
    return function(command) {
      return consoleInterface.performAction(command, action);
    };
  };

  ConsoleInterface.prototype.performAction = function(command, action) {
    var instanceConfig,
      _this = this;
    instanceConfig = this.extractConfig(command);
    this.docpad.action('load ready', instanceConfig, function(err) {
      if (err) {
        return _this.completeAction(err);
      }
      return action(_this.completeAction);
    });
    return this;
  };

  ConsoleInterface.prototype.completeAction = function(err) {
    var docpad, locale;
    docpad = this.docpad;
    locale = docpad.getLocale();
    if (err) {
      this.handleError(err);
    } else {
      docpad.log('info', locale.consoleSuccess);
    }
    return this;
  };

  ConsoleInterface.prototype.extractConfig = function(customConfig) {
    var commanderConfig, config, configPath, key, outPath, sourceConfig, value;
    if (customConfig == null) {
      customConfig = {};
    }
    config = {};
    commanderConfig = this.commander;
    sourceConfig = this.docpad.initialConfig;
    if (commanderConfig.debug) {
      if (commanderConfig.debug === true) {
        commanderConfig.debug = 7;
      }
      commanderConfig.logLevel = commanderConfig.debug;
    }
    if (commanderConfig.config) {
      configPath = pathUtil.resolve(process.cwd(), commanderConfig.config);
      commanderConfig.configPaths = [configPath];
    }
    if (commanderConfig.out) {
      outPath = pathUtil.resolve(process.cwd(), commanderConfig.out);
      commanderConfig.outPath = outPath;
    }
    for (key in commanderConfig) {
      if (!__hasProp.call(commanderConfig, key)) continue;
      value = commanderConfig[key];
      if (typeof sourceConfig[key] !== 'undefined') {
        config[key] = value;
      }
    }
    for (key in customConfig) {
      if (!__hasProp.call(customConfig, key)) continue;
      value = customConfig[key];
      if (typeof sourceConfig[key] !== 'undefined') {
        config[key] = value;
      }
    }
    return config;
  };

  ConsoleInterface.prototype.selectSkeletonCallback = function(skeletonsCollection, next) {
    var commander, docpad, locale, skeletonNames;
    commander = this.commander;
    docpad = this.docpad;
    locale = docpad.getLocale();
    skeletonNames = [];
    console.log(cliColor.bold(locale.skeletonSelectionIntroduction + '\n'));
    skeletonsCollection.forEach(function(skeletonModel) {
      var skeletonDescription, skeletonName;
      skeletonName = skeletonModel.get('name');
      skeletonDescription = skeletonModel.get('description').replace(/\n/g, '\n\t');
      skeletonNames.push(skeletonName);
      return console.log("\t" + (cliColor.bold(skeletonName)) + "\n\t" + skeletonDescription + "\n");
    });
    console.log(cliColor.bold(locale.skeletonSelectionPrompt));
    commander.choose(skeletonNames, function(i) {
      process.stdin.destroy();
      return next(null, skeletonsCollection.at(i));
    });
    return this;
  };

  ConsoleInterface.prototype.welcomeCallback = function(opts, next) {
    var balUtil, commander, consoleInterface, docpad, locale, userConfig, welcomeTasks;
    balUtil = require('bal-util');
    consoleInterface = this;
    commander = this.commander;
    docpad = this.docpad;
    locale = docpad.getLocale();
    userConfig = docpad.userConfig;
    welcomeTasks = new balUtil.Group(next);
    welcomeTasks.push(function(complete) {
      if (docpad.config.prompts === false || userConfig.tos === true) {
        return complete();
      }
      return consoleInterface.confirm(locale.tosPrompt, true, function(ok) {
        return docpad.track('tos', {
          ok: ok
        }, function(err) {
          if (ok) {
            userConfig.tos = true;
            console.log(locale.tosAgree);
            docpad.updateUserConfig(complete);
          } else {
            console.log(locale.tosDisagree);
            process.exit();
          }
        });
      });
    });
    welcomeTasks.push(function(complete) {
      if (docpad.config.prompts === false || (userConfig.subscribed != null) || ((userConfig.subscribeTryAgain != null) && (new Date()) > (new Date(userConfig.subscribeTryAgain)))) {
        return complete();
      }
      return consoleInterface.confirm(locale.subscribePrompt, true, function(ok) {
        return docpad.track('subscribe', {
          ok: ok
        }, function(err) {
          var commands;
          if (!ok) {
            console.log(locale.subscribeIgnore);
            userConfig.subscribed = false;
            docpad.updateUserConfig(function(err) {
              if (err) {
                return complete(err);
              }
              return balUtil.wait(5000, complete);
            });
            return;
          }
          commands = [['config', '--get', 'user.name'], ['config', '--get', 'user.email'], ['config', '--get', 'github.user']];
          return balUtil.gitCommands(commands, function(err, results) {
            var subscribeTasks, _ref, _ref1, _ref2;
            userConfig.name || (userConfig.name = String((results != null ? (_ref = results[0]) != null ? _ref[1] : void 0 : void 0) || '').trim() || null);
            userConfig.email || (userConfig.email = String((results != null ? (_ref1 = results[1]) != null ? _ref1[1] : void 0 : void 0) || '').trim() || null);
            userConfig.username || (userConfig.username = String((results != null ? (_ref2 = results[2]) != null ? _ref2[1] : void 0 : void 0) || '').trim() || null);
            if (userConfig.name || userConfig.email || userConfig.username) {
              console.log(locale.subscribeConfigNotify);
            }
            subscribeTasks = new balUtil.Group(function(err) {
              if (err) {
                console.log(locale.subscribeError);
                userConfig.subscribeTryAgain = new Date().getTime() + 1000 * 60 * 60 * 24;
              } else {
                console.log(locale.subscribeSuccess);
                userConfig.subscribed = true;
                userConfig.subscribeTryAgain = null;
              }
              return docpad.updateUserConfig(userConfig, complete);
            });
            subscribeTasks.push(function(complete) {
              return consoleInterface.prompt(locale.subscribeNamePrompt, userConfig.name, function(result) {
                userConfig.name = result;
                return complete();
              });
            });
            subscribeTasks.push(function(complete) {
              return consoleInterface.prompt(locale.subscribeEmailPrompt, userConfig.email, function(result) {
                userConfig.email = result;
                return complete();
              });
            });
            subscribeTasks.push(function(complete) {
              return consoleInterface.prompt(locale.subscribeUsernamePrompt, userConfig.username, function(result) {
                userConfig.username = result;
                return complete();
              });
            });
            subscribeTasks.push(function(complete) {
              return docpad.updateUserConfig(complete);
            });
            subscribeTasks.push(function(complete) {
              var requestUrl;
              console.log(locale.subscribeProgress);
              requestUrl = docpad.config.helperUrl + '?' + require('querystring').stringify({
                method: 'add-subscriber',
                name: userConfig.name,
                email: userConfig.email,
                username: userConfig.username
              });
              return balUtil.readPath(requestUrl, function(err, body) {
                var data;
                if (err) {
                  docpad.log('debug', locale.subscribeRequestError, err.message);
                  return complete(err);
                }
                docpad.log('debug', locale.subscribeRequestData, body);
                try {
                  data = JSON.parse(body);
                  if (data.success === false) {
                    return complete(new Error(data.error || 'unknown error'));
                  } else {
                    return complete();
                  }
                } catch (err) {
                  return complete(err);
                }
              });
            });
            return subscribeTasks.sync();
          });
        });
      });
    });
    welcomeTasks.sync();
    return this;
  };

  ConsoleInterface.prototype.prompt = function(message, fallback, next) {
    var commander, consoleInterface;
    consoleInterface = this;
    commander = this.commander;
    if (fallback) {
      message += " [" + fallback + "]";
    }
    commander.prompt(message + ' ', function(result) {
      if (!result.trim()) {
        if (fallback != null) {
          result = fallback;
        } else {
          return consoleInterface.prompt(message, fallback, next);
        }
      }
      return next(result);
    });
    return this;
  };

  ConsoleInterface.prototype.confirm = function(message, fallback, next) {
    var commander, consoleInterface;
    consoleInterface = this;
    commander = this.commander;
    if (fallback === true) {
      message += " [Y/n]";
    } else if (fallback === false) {
      message += " [y/N]";
    }
    commander.prompt(message + ' ', function(ok) {
      if (!ok.trim()) {
        if (fallback != null) {
          ok = fallback;
        } else {
          return consoleInterface.confirm(message, fallback, next);
        }
      } else {
        ok = /^y|yes|ok|true$/i.test(ok);
      }
      return next(ok);
    });
    return this;
  };

  ConsoleInterface.prototype.generate = function(next) {
    this.docpad.action('generate', next);
    return this;
  };

  ConsoleInterface.prototype.help = function(next) {
    var help;
    help = this.commander.helpInformation();
    console.log(help);
    next();
    return this;
  };

  ConsoleInterface.prototype.info = function(next) {
    var info;
    info = require('util').inspect(this.docpad.config);
    console.log(info);
    next();
    return this;
  };

  ConsoleInterface.prototype.install = function(next) {
    this.docpad.action('install', next);
    return this;
  };

  ConsoleInterface.prototype.render = function(next) {
    var balUtil, basename, commander, data, docpad, filename, opts, renderDocument, stdin, timeout, useStdin;
    docpad = this.docpad;
    commander = this.commander;
    balUtil = require('bal-util');
    opts = {};
    filename = commander.args[0] || null;
    basename = pathUtil.basename(filename);
    opts.filename = filename;
    opts.renderSingleExtensions = 'auto';
    data = '';
    useStdin = true;
    renderDocument = function() {
      return docpad.action('render', opts, function(err, result) {
        if (err) {
          return docpad.fatal(err);
        }
        if (commander.out != null) {
          return balUtil.writeFile(commander.out, result, next);
        } else {
          process.stdout.write(result);
          return next();
        }
      });
    };
    timeout = setTimeout(function() {
      timeout = null;
      if (data.replace(/\s+/, '')) {
        return;
      }
      useStdin = false;
      stdin.pause();
      return renderDocument();
    }, 1000);
    stdin = process.stdin;
    stdin.resume();
    stdin.setEncoding('utf8');
    stdin.on('data', function(_data) {
      return data += _data.toString();
    });
    process.stdin.on('end', function() {
      if (!useStdin) {
        return;
      }
      if (timeout) {
        clearTimeout(timeout);
        timeout = null;
      }
      opts.data = data;
      return renderDocument();
    });
    return this;
  };

  ConsoleInterface.prototype.run = function(next) {
    this.docpad.action('run', {
      selectSkeletonCallback: this.selectSkeletonCallback
    }, next);
    return this;
  };

  ConsoleInterface.prototype.server = function(next) {
    this.docpad.action('server generate', next);
    return this;
  };

  ConsoleInterface.prototype.clean = function(next) {
    this.docpad.action('clean', next);
    return this;
  };

  ConsoleInterface.prototype.skeleton = function(next) {
    this.docpad.action('skeleton', {
      selectSkeletonCallback: this.selectSkeletonCallback
    }, next);
    return this;
  };

  ConsoleInterface.prototype.watch = function(next) {
    this.docpad.action('generate watch', next);
    return this;
  };

  return ConsoleInterface;

})();

module.exports = ConsoleInterface;
