// docpad/out/lib/plugin-loader.js
// Generated by CoffeeScript 1.6.3
var PluginLoader, pathUtil, safefs, semver, util,
  __indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

pathUtil = require('path');

semver = require('semver');

safefs = require('safefs');

util = require('util');

PluginLoader = (function() {
  PluginLoader.prototype.docpad = null;

  PluginLoader.prototype.BasePlugin = null;

  PluginLoader.prototype.dirPath = null;

  PluginLoader.prototype.packagePath = null;

  PluginLoader.prototype.packageData = {};

  PluginLoader.prototype.pluginPath = null;

  PluginLoader.prototype.pluginClass = {};

  PluginLoader.prototype.pluginName = null;

  PluginLoader.prototype.pluginVersion = null;

  PluginLoader.prototype.nodeModulesPath = null;

  function PluginLoader(_arg) {
    this.docpad = _arg.docpad, this.dirPath = _arg.dirPath, this.BasePlugin = _arg.BasePlugin;
    docpad = this.docpad;
    this.pluginName = pathUtil.basename(this.dirPath).replace(/^docpad-plugin-/, '');
    this.pluginClass = {};
    this.packageData = {};
    this.nodeModulesPath = pathUtil.resolve(this.dirPath, 'node_modules');
  }

  PluginLoader.prototype.exists = function(next) {
    var failure, packagePath, success,
      _this = this;
    packagePath = this.packagePath || pathUtil.resolve(this.dirPath, "package.json");
    failure = function(err) {
      if (err == null) {
        err = null;
      }
      return next(err, false);
    };
    success = function() {
      return next(null, true);
    };
    safefs.exists(packagePath, function(exists) {
      if (!exists) {
        return failure();
      }
      _this.packagePath = packagePath;
      return safefs.readFile(packagePath, function(err, data) {
        var pluginPath, pluginVersion;
        if (err) {
          return failure(err);
        }
        try {
          _this.packageData = JSON.parse(data.toString());
        } catch (_error) {
          err = _error;
          return failure(err);
        } finally {
          if (!_this.packageData) {
            return failure();
          }
        }
        pluginVersion = _this.packageData.version;
        pluginPath = _this.packageData.main && pathUtil.join(_this.dirPath, _this.packageData.main);
        if (!pluginVersion) {
          return failure();
        }
        if (!pluginPath) {
          return failure();
        }
        _this.pluginVersion = pluginVersion;
        _this.pluginPath = pluginPath;
        return success();
      });
    });
    return this;
  };

  PluginLoader.prototype.unsupported = function(next) {
    var docpad, engines, keywords, peerDependencies, platforms, unsupported, version, _ref;
    docpad = this.docpad;
    version = this.packageData.version;
    keywords = this.packageData.keywords || [];
    platforms = this.packageData.platforms || [];
    engines = this.packageData.engines || {};
    peerDependencies = this.packageData.peerDependencies || {};
    unsupported = __indexOf.call(keywords, 'docpad-plugin') < 0 ? 'type' : version && !semver.satisfies(version, docpad.pluginVersion) ? 'version-plugin' : platforms.length && (_ref = process.platform, __indexOf.call(platforms, _ref) < 0) ? 'platform' : (engines.node != null) && !semver.satisfies(process.version, engines.node) ? 'engine-node' : (engines.docpad != null) && !semver.satisfies(docpad.version, engines.docpad) ? 'version-docpad' : (peerDependencies.docpad != null) && !semver.satisfies(docpad.version, peerDependencies.docpad) ? 'version-docpad' : false;
    next(null, unsupported);
    return this;
  };

  PluginLoader.prototype.install = function(next) {
    var docpad;
    docpad = this.docpad;
    if (this.packagePath) {
      docpad.initNodeModules({
        path: this.dirPath,
        next: function(err, results) {
          return next(err);
        }
      });
    } else {
      next();
    }
    return this;
  };

  PluginLoader.prototype.load = function(next) {
    var docpad, err, locale, pluginPrototypeName, validPluginName, _base;
    docpad = this.docpad;
    locale = docpad.getLocale();
    try {
      this.pluginClass = require(this.pluginPath)(this.BasePlugin);
      if ((_base = this.pluginClass.prototype).version == null) {
        _base.version = this.pluginVersion;
      }
      pluginPrototypeName = this.pluginClass.prototype.name;
      if (/^[a-z0-9]+$/.test(this.pluginName) === false) {
        validPluginName = this.pluginName.replace(/[^a-z0-9]/, '');
        docpad.log('warn', util.format(locale.pluginNamingConventionInvalid, this.pluginName, validPluginName));
      }
      if (pluginPrototypeName === null) {
        this.pluginClass.prototype.name = this.pluginName;
        docpad.log('warn', util.format(locale.pluginPrototypeNameUndefined, this.pluginName));
      } else if (pluginPrototypeName !== this.pluginName) {
        docpad.log('warn', util.format(locale.pluginPrototypeNameDifferent, this.pluginName, pluginPrototypeName));
      }
    } catch (_error) {
      err = _error;
      return next(err, null);
    }
    next(null, this.pluginClass);
    return this;
  };

  PluginLoader.prototype.create = function(config, next) {
    var docpad, err, pluginInstance;
    try {
      docpad = this.docpad;
      pluginInstance = new this.pluginClass({
        docpad: docpad,
        config: config
      });
    } catch (_error) {
      err = _error;
      return next(err, null);
    }
    return next(null, pluginInstance);
    return this;
  };

  return PluginLoader;

})();

module.exports = PluginLoader;
