// docpad/out/lib/plugin-loader.js
// Generated by CoffeeScript 1.4.0
var PluginLoader, balUtil, pathUtil, semver, util,
  __indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

pathUtil = require('path');

semver = require('semver');

balUtil = require('bal-util');

util = require('util');

PluginLoader = (function() {

  PluginLoader.prototype.docpad = null;

  PluginLoader.prototype.BasePlugin = null;

  PluginLoader.prototype.dirPath = null;

  PluginLoader.prototype.packagePath = null;

  PluginLoader.prototype.packageData = {};

  PluginLoader.prototype.pluginPath = null;

  PluginLoader.prototype.pluginClass = {};

  PluginLoader.prototype.pluginName = null;

  PluginLoader.prototype.pluginVersion = null;

  PluginLoader.prototype.nodeModulesPath = null;

  function PluginLoader(_arg) {
    this.docpad = _arg.docpad, this.dirPath = _arg.dirPath, this.BasePlugin = _arg.BasePlugin;
    docpad = this.docpad;
    this.pluginName = pathUtil.basename(this.dirPath).replace(/^docpad-plugin-/, '');
    this.pluginClass = {};
    this.packageData = {};
    this.nodeModulesPath = pathUtil.resolve(this.dirPath, 'node_modules');
  }

  PluginLoader.prototype.exists = function(next) {
    var packagePath, pluginPath,
      _this = this;
    packagePath = this.packagePath || pathUtil.resolve(this.dirPath, "package.json");
    pluginPath = this.pluginPath || pathUtil.resolve(this.dirPath, "" + this.pluginName + ".plugin.coffee");
    balUtil.exists(packagePath, function(exists) {
      if (!exists) {
        return balUtil.exists(pluginPath, function(exists) {
          if (!exists) {
            return next(null, false);
          } else {
            _this.pluginPath = pluginPath;
            return next(null, true);
          }
        });
      } else {
        _this.packagePath = packagePath;
        return balUtil.readFile(packagePath, function(err, data) {
          if (err) {
            return next(err, false);
          } else {
            try {
              _this.packageData = JSON.parse(data.toString());
            } catch (err) {
              return next(err, false);
            }
            if (!_this.packageData) {
              return next(null, false);
            }
            pluginPath = (_this.packageData.main != null) && pathUtil.join(_this.dirPath, _this.pluginPath) || pluginPath;
            _this.pluginVersion = _this.packageData.version;
            return balUtil.exists(pluginPath, function(exists) {
              if (!exists) {
                return next(null, false);
              } else {
                _this.pluginPath = pluginPath;
                return next(null, true);
              }
            });
          }
        });
      }
    });
    return this;
  };

  PluginLoader.prototype.unsupported = function(next) {
    var docpad, engines, keywords, platforms, unsupported, _ref;
    docpad = this.docpad;
    unsupported = false;
    if (this.packageData) {
      keywords = this.packageData.keywords || [];
      if (__indexOf.call(keywords, 'docpad-plugin') < 0) {
        unsupported = 'type';
      }
    }
    if (this.packageData && this.packageData.platforms) {
      platforms = this.packageData.platforms || [];
      if (_ref = process.platform, __indexOf.call(platforms, _ref) < 0) {
        unsupported = 'platform';
      }
    }
    if (this.packageData && this.packageData.engines) {
      engines = this.packageData.engines || {};
      if (engines.node != null) {
        if (!semver.satisfies(process.version, engines.node)) {
          unsupported = 'engine';
        }
      }
      if (engines.docpad != null) {
        if (!semver.satisfies(docpad.version, engines.docpad)) {
          unsupported = 'version';
        }
      }
    }
    next(null, unsupported);
    return this;
  };

  PluginLoader.prototype.install = function(next) {
    var docpad;
    docpad = this.docpad;
    if (this.packagePath) {
      docpad.initNodeModules({
        path: this.dirPath,
        next: function(err, results) {
          return next(err);
        }
      });
    } else {
      next();
    }
    return this;
  };

  PluginLoader.prototype.load = function(next) {
    var docpad, locale, pluginPrototypeName, validPluginName, _base, _ref;
    docpad = this.docpad;
    locale = docpad.getLocale();
    try {
      this.pluginClass = require(this.pluginPath)(this.BasePlugin);
      if ((_ref = (_base = this.pluginClass.prototype).version) == null) {
        _base.version = this.pluginVersion;
      }
      pluginPrototypeName = this.pluginClass.prototype.name;
      if (/^[a-z0-9]+$/.test(this.pluginName) === false) {
        validPluginName = this.pluginName.replace(/[^a-z0-9]/, '');
        docpad.log('warn', util.format(locale.pluginNamingConventionInvalid, this.pluginName, validPluginName));
      }
      if (pluginPrototypeName === null) {
        this.pluginClass.prototype.name = this.pluginName;
        docpad.log('warn', util.format(locale.pluginPrototypeNameUndefined, this.pluginName));
      } else if (pluginPrototypeName !== this.pluginName) {
        docpad.log('warn', util.format(locale.pluginPrototypeNameDifferent, this.pluginName, pluginPrototypeName));
      }
    } catch (err) {
      return next(err, null);
    }
    next(null, this.pluginClass);
    return this;
  };

  PluginLoader.prototype.create = function(userConfiguration, next) {
    var config, docpad, pluginInstance;
    if (userConfiguration == null) {
      userConfiguration = {};
    }
    try {
      config = balUtil.deepExtendPlainObjects({}, this.docpad.config.plugins[this.pluginName], userConfiguration);
      docpad = this.docpad;
      pluginInstance = new this.pluginClass({
        docpad: docpad,
        config: config
      });
    } catch (err) {
      return next(err, null);
    }
    return next(null, pluginInstance);
    return this;
  };

  return PluginLoader;

})();

module.exports = PluginLoader;
